- DataFrame：資料庫
 - 搜尋
 - 清理資料
 - 統計資料
- 介面：
 - CLI：最好做，但畫面不好看
 - Streamlit：share data apps分享資料的介面
 - gradio：與自然語言模型搭配的介面
 - flask：網站介面（小型）
 - django：網站介面
 - plotly：資料圖像呈現介面
 - tkinter：window內建的視窗介面
 - render

In [5]:
import pandas as pd

df = pd.read_csv('各鄉鎮市區人口密度.csv') #把csv檔讀成dataframe
df

statistic_yyy                                            site_id  \
0             統計年                                                區域別   
1             106                                             新北市板橋區   
2             106                                             新北市三重區   
3             106                                             新北市中和區   
4             106                                             新北市永和區   
..            ...                                                ...   
371           NaN                                                NaN   
372           NaN                           說明：1.人口密度係指每單位土地面積內之人口數。   
373           NaN  2.96年12月起，我國土地面積增列東沙群島(2.38平方公里)及南沙群島(0.4896平方公里)   
374           NaN           ，由高雄市代管；原金門縣烏坵鄉面積，因重測由2.6平方公里修正為1.2平方公里。   
375           NaN           3.104年12月南沙太平島交通基礎整建工程竣工，增加面積0.0149平方公里。   

    people_total     area population_density  
0          年底人口數     土地面積               人口密度  
1         551480  23.1373              23835  
2         387484   16.317              23747  
3         413590   20.144              20532  
4         222585   5.7138              38956  
..           ...      ...                ...  
371          NaN      NaN                NaN  
372          NaN      NaN                NaN  
373          NaN      NaN                NaN  
374          NaN      NaN                NaN  
375          NaN      NaN                NaN  

[376 rows x 5 columns]

In [ ]:
df.info() #資料裡有什麼

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 376 entries, 0 to 375
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   statistic_yyy       371 non-null    object
 1   site_id             375 non-null    object
 2   people_total        371 non-null    object
 3   area                371 non-null    object
 4   population_density  365 non-null    object
dtypes: object(5)
memory usage: 14.8+ KB


- RangeIndex：index（索引）有376個（0~375）
- 5個columns（欄位），第0欄有371筆不是null（也就是有5筆是null）
- 類型（Dtype）是字串

In [4]:
#更改columns名稱
df1 = df.rename(columns={'statistic_yyy':'統計年',
                        'site_id':'區域別',
                        'people_total':'年底人口數',
                        'area':'土地面積',
                        'population_density':'人口密度'})
df1

統計年                                                區域別   年底人口數     土地面積  \
0    統計年                                                區域別   年底人口數     土地面積   
1    106                                             新北市板橋區  551480  23.1373   
2    106                                             新北市三重區  387484   16.317   
3    106                                             新北市中和區  413590   20.144   
4    106                                             新北市永和區  222585   5.7138   
..   ...                                                ...     ...      ...   
371  NaN                                                NaN     NaN      NaN   
372  NaN                           說明：1.人口密度係指每單位土地面積內之人口數。     NaN      NaN   
373  NaN  2.96年12月起，我國土地面積增列東沙群島(2.38平方公里)及南沙群島(0.4896平方公里)     NaN      NaN   
374  NaN           ，由高雄市代管；原金門縣烏坵鄉面積，因重測由2.6平方公里修正為1.2平方公里。     NaN      NaN   
375  NaN           3.104年12月南沙太平島交通基礎整建工程竣工，增加面積0.0149平方公里。     NaN      NaN   

      人口密度  
0     人口密度  
1    23835  
2    23747  
3    20532  
4    38956  
..     ...  
371    NaN  
372    NaN  
373    NaN  
374    NaN  
375    NaN  

[376 rows x 5 columns]

In [12]:
#刪除第0列（index 0） 
df2 = df1.drop(index=0) #預設axis=0，axis=0是index，axis=1是column
#刪除有NaN的列
df3 = df2.dropna(axis=0,how='any')
#刪除欄
df4 = df3.drop('統計年',axis=1)
df4


區域別   年底人口數     土地面積   人口密度
1    新北市板橋區  551480  23.1373  23835
2    新北市三重區  387484   16.317  23747
3    新北市中和區  413590   20.144  20532
4    新北市永和區  222585   5.7138  38956
5    新北市新莊區  416524  19.7383  21102
..      ...     ...      ...    ...
364  金門縣烏坵鄉     685      1.2    571
365  連江縣南竿鄉    7544     10.4    725
366  連江縣北竿鄉    2360      9.9    238
369    東沙群島      …      2.38     … 
370    南沙群島      …    0.5045     … 

[364 rows x 4 columns]

In [13]:
df4.info()

<class 'pandas.core.frame.DataFrame'>
Index: 364 entries, 1 to 370
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   區域別     364 non-null    object
 1   年底人口數   364 non-null    object
 2   土地面積    364 non-null    object
 3   人口密度    364 non-null    object
dtypes: object(4)
memory usage: 14.2+ KB


```
lambda x : x*2
```
- lambda代表匿名的function()，x代表參數名稱，冒號後是程式區塊，會自動return
- lambda只能寫一行

**用apply才能使用外部function**

In [ ]:
#單一欄位是series，要更改單一欄位的資料，要找series的實體方法
#因為人口數中不是每一個都數值，所以用function一個一個處理
def strTOint(x:str)->int:
    try:
        return int(x) #把傳進來的值轉成int數值
    except:
        return 0 #失敗的話，轉為0

df4['年底人口數'] = df4['年底人口數'].apply(strTOint) #apply：使用這個function
df4['人口密度'] = df4['人口密度'].apply(strTOint)

#可以全部一起改的就直接用series的實體方法
df4['土地面積'] = df4['土地面積'].astype('float')

df4.info()


<class 'pandas.core.frame.DataFrame'>
Index: 364 entries, 1 to 370
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   區域別     364 non-null    object 
 1   年底人口數   364 non-null    int64  
 2   土地面積    364 non-null    float64
 3   人口密度    364 non-null    int64  
dtypes: float64(1), int64(2), object(1)
memory usage: 14.2+ KB


In [17]:
#抓資料做新欄位

def split_area(v:str)->str:
    if len(v) >= 6: #如果字數>=6
        return v[:3] #抓前3個字
    else:
        return v #不然直接回傳地區

def split_area1(v:str)->str:
    if len(v) >=6:
        return v[-3:] #抓後3個字
    else:
        return v

df4['縣市'] = df4['區域別'].apply(split_area)
df4['鄉鎮市區'] = df4['區域別'].apply(split_area1)
df4

區域別   年底人口數     土地面積   人口密度    縣市  鄉鎮市區
1    新北市板橋區  551480  23.1373  23835   新北市   板橋區
2    新北市三重區  387484  16.3170  23747   新北市   三重區
3    新北市中和區  413590  20.1440  20532   新北市   中和區
4    新北市永和區  222585   5.7138  38956   新北市   永和區
5    新北市新莊區  416524  19.7383  21102   新北市   新莊區
..      ...     ...      ...    ...   ...   ...
364  金門縣烏坵鄉     685   1.2000    571   金門縣   烏坵鄉
365  連江縣南竿鄉    7544  10.4000    725   連江縣   南竿鄉
366  連江縣北竿鄉    2360   9.9000    238   連江縣   北竿鄉
369    東沙群島       0   2.3800      0  東沙群島  東沙群島
370    南沙群島       0   0.5045      0  南沙群島  南沙群島

[364 rows x 6 columns]

In [20]:
#刪除欄位並重新排列
df5 = df4.reindex(columns=['縣市','鄉鎮市區','年底人口數','土地面積','人口密度'])
df5

縣市  鄉鎮市區   年底人口數     土地面積   人口密度
1     新北市   板橋區  551480  23.1373  23835
2     新北市   三重區  387484  16.3170  23747
3     新北市   中和區  413590  20.1440  20532
4     新北市   永和區  222585   5.7138  38956
5     新北市   新莊區  416524  19.7383  21102
..    ...   ...     ...      ...    ...
364   金門縣   烏坵鄉     685   1.2000    571
365   連江縣   南竿鄉    7544  10.4000    725
366   連江縣   北竿鄉    2360   9.9000    238
369  東沙群島  東沙群島       0   2.3800      0
370  南沙群島  南沙群島       0   0.5045      0

[364 rows x 5 columns]

In [21]:
#搜尋
df5.query('年底人口數>=300000')

縣市 鄉鎮市區   年底人口數      土地面積   人口密度
1    新北市  板橋區  551480   23.1373  23835
2    新北市  三重區  387484   16.3170  23747
3    新北市  中和區  413590   20.1440  20532
5    新北市  新莊區  416524   19.7383  21102
6    新北市  新店區  302089  120.2255   2513
32   臺北市  大安區  309969   11.3614  27283
42   桃園市  桃園區  440840   34.8046  12666
125  高雄市  三民區  343243   19.7866  17347
132  高雄市  鳳山區  359120   26.7590  13421

In [ ]:
#最大值
m = df5['人口密度'].max() #最大值放入變數m
df5.query('人口密度 == @m') #在query裡呼叫變數，前面加@

縣市 鄉鎮市區   年底人口數    土地面積   人口密度
4  新北市  永和區  222585  5.7138  38956

In [25]:
#把縣市一樣的合起來，計算各項平均
df5.groupby(by='縣市')[['年底人口數','土地面積','人口密度']].mean() 

年底人口數        土地面積          人口密度
縣市                                            
南投縣     38542.384615  315.879692    338.692308
南沙群島        0.000000    0.504500      0.000000
嘉義市東區  122103.000000   30.155600   4049.000000
嘉義市西區  147295.000000   29.870000   4931.000000
嘉義縣     28399.000000  105.757594    383.666667
基隆市     53065.428571   18.965557   4603.857143
宜蘭縣     38050.583333  178.635425   1125.416667
屏東縣     25149.666667   84.109100    567.666667
彰化縣     49325.307692   41.322923   1293.307692
新北市    147246.296296   61.485107   7190.629630
新竹市     77943.000000   54.849100   1421.000000
新竹市北區  150636.000000   15.726700   9578.000000
新竹市東區  212553.000000   33.576800   6330.000000
新竹縣     45552.583333  115.061325    839.750000
東沙群島        0.000000    2.380000      0.000000
桃園市    148566.750000   95.369500   2863.000000
澎湖縣     17345.500000   21.144017    689.166667
臺中市     96019.250000   91.053296   2328.083333
臺中市中區   18450.000000    0.880300  20959.000000
臺中市北區  147653.000000    6.937600  21283.000000
臺中市南區  125067.000000    6.810100  18365.000000
臺中市東區   75811.000000    9.285500   8164.000000
臺中市西區  115627.000000    5.704200  20271.000000
臺北市    223604.750000   22.649975  15377.666667
臺南市     42398.852941   62.956924   1264.911765
臺南市北區  132726.000000   10.434000  12721.000000
臺南市南區  125394.000000   27.268100   4599.000000
臺南市東區  186841.000000   13.415600  13927.000000
臺東縣     13721.250000  219.703287    133.375000
花蓮縣     25325.923077  356.043954    466.692308
苗栗縣     30767.055556  101.128606    571.222222
連江縣      4952.000000   10.150000    481.500000
金門縣     22909.333333   25.276000    932.666667
雲林縣     34518.650000   64.541630    560.050000
高雄市     73076.631579   77.680326   5116.052632

In [ ]:
#用agg同時顯示多個統計量，也可以改顯示名稱
df6 = df5.groupby(by='縣市')[['年底人口數','土地面積']].agg([('平均','mean'),('加總','sum'),'max','min']) 

#加上欄位名稱
df6.columns.names = ["資料", "統計"]

df6

資料             年底人口數                                 土地面積             \
統計                平均       加總     max     min          平均         加總   
縣市                                                                     
南投縣     38542.384615   501051  100180   10904  315.879692  4106.4360   
南沙群島        0.000000        0       0       0    0.504500     0.5045   
嘉義市東區  122103.000000   122103  122103  122103   30.155600    30.1556   
嘉義市西區  147295.000000   147295  147295  147295   29.870000    29.8700   
嘉義縣     28399.000000   511182   71693    4564  105.757594  1903.6367   
基隆市     53065.428571   371458   82358   38708   18.965557   132.7589   
宜蘭縣     38050.583333   456607   95918    5900  178.635425  2143.6251   
屏東縣     25149.666667   829939  200452    3246   84.109100  2775.6003   
彰化縣     49325.307692  1282458  233613   15303   41.322923  1074.3960   
新北市    147246.296296  3975650  551480    6528   61.485107  1660.0979   
新竹市     77943.000000    77943   77943   77943   54.849100    54.8491   
新竹市北區  150636.000000   150636  150636  150636   15.726700    15.7267   
新竹市東區  212553.000000   212553  212553  212553   33.576800    33.5768   
新竹縣     45552.583333   546631  181955    4559  115.061325  1380.7359   
東沙群島        0.000000        0       0       0    2.380000     2.3800   
桃園市    148566.750000  1782801  440840   11505   95.369500  1144.4340   
澎湖縣     17345.500000   104073   62308    3856   21.144017   126.8641   
臺中市     96019.250000  2304462  274819   10949   91.053296  2185.2791   
臺中市中區   18450.000000    18450   18450   18450    0.880300     0.8803   
臺中市北區  147653.000000   147653  147653  147653    6.937600     6.9376   
臺中市南區  125067.000000   125067  125067  125067    6.810100     6.8101   
臺中市東區   75811.000000    75811   75811   75811    9.285500     9.2855   
臺中市西區  115627.000000   115627  115627  115627    5.704200     5.7042   
臺北市    223604.750000  2683257  309969  122155   22.649975   271.7997   
臺南市     42398.852941  1441561  233905    4038   62.956924  2140.5354   
臺南市北區  132726.000000   132726  132726  132726   10.434000    10.4340   
臺南市南區  125394.000000   125394  125394  125394   27.268100    27.2681   
臺南市東區  186841.000000   186841  186841  186841   13.415600    13.4156   
臺東縣     13721.250000   219540  105960    3518  219.703287  3515.2526   
花蓮縣     25325.923077   329237  104380    4383  356.043954  4628.5714   
苗栗縣     30767.055556   553807  103147    4417  101.128606  1820.3149   
連江縣      4952.000000     9904    7544    2360   10.150000    20.3000   
金門縣     22909.333333   137456   43285     685   25.276000   151.6560   
雲林縣     34518.650000   690373  108567   13010   64.541630  1290.8326   
高雄市     73076.631579  2776912  359120    1924   77.680326  2951.8524   

資料                         
統計           max      min  
縣市                         
南投縣    1422.4188  49.7268  
南沙群島      0.5045   0.5045  
嘉義市東區    30.1556  30.1556  
嘉義市西區    29.8700  29.8700  
嘉義縣     427.8471  33.0463  
基隆市      56.2659   4.2335  
宜蘭縣     740.6520  11.3448  
屏東縣     301.0018   6.8018  
彰化縣      92.8478  18.0856  
新北市     191.4508   5.7138  
新竹市      54.8491  54.8491  
新竹市北區    15.7267  15.7267  
新竹市東區    33.5768  33.5768  
新竹縣     527.5795  40.7858  
東沙群島      2.3800   2.3800  
桃園市     350.7775  33.7111  
澎湖縣      33.9918   6.9868  
臺中市    1037.8192  16.6049  
臺中市中區     0.8803   0.8803  
臺中市北區     6.9376   6.9376  
臺中市南區     6.8101   6.8101  
臺中市東區     9.2855   9.2855  
臺中市西區     5.7042   5.7042  
臺北市      62.3682   5.6815  
臺南市     171.5198   6.2600  
臺南市北區    10.4340  10.4340  
臺南市南區    27.2681  27.2681  
臺南市東區    13.4156  13.4156  
臺東縣     880.0382  15.0919  
花蓮縣    1641.8555  29.4095  
苗栗縣     614.5127  37.5592  
連江縣      10.4000   9.9000  
金門縣      41.6960   1.2000  
雲林縣     166.6059  37.0552  
高雄市     928.9800   1.4161

In [ ]:
#變series
df6.stack(level=['資料','統計'], future_stack=True)

縣市   資料     統計 
南投縣  年底人口數  平均      38542.384615
            加總     501051.000000
            max    100180.000000
            min     10904.000000
     土地面積   平均        315.879692
                       ...      
高雄市  年底人口數  min      1924.000000
     土地面積   平均         77.680326
            加總       2951.852400
            max       928.980000
            min         1.416100
Length: 280, dtype: float64

In [ ]:
#改欄位，改變欄和列 
df6.stack(level=['資料','統計'], future_stack=True).unstack(level=['縣市','統計'])

縣市              南投縣                                         南沙群島          \
統計               平均          加總          max         min      平均      加總   
資料                                                                         
年底人口數  38542.384615  501051.000  100180.0000  10904.0000  0.0000  0.0000   
土地面積     315.879692    4106.436    1422.4188     49.7268  0.5045  0.5045   

縣市                           嘉義市東區               ...        金門縣         \
統計        max     min           平均           加總  ...        max    min   
資料                                               ...                     
年底人口數  0.0000  0.0000  122103.0000  122103.0000  ...  43285.000  685.0   
土地面積   0.5045  0.5045      30.1556      30.1556  ...     41.696    1.2   

縣市             雲林縣                                                 高雄市  \
統計              平均           加總          max         min            平均   
資料                                                                       
年底人口數  34518.65000  690373.0000  108567.0000  13010.0000  73076.631579   
土地面積      64.54163    1290.8326     166.6059     37.0552     77.680326   

縣市                                         
統計               加總        max        min  
資料                                         
年底人口數  2.776912e+06  359120.00  1924.0000  
土地面積   2.951852e+03     928.98     1.4161  

[2 rows x 140 columns]

In [33]:
import numpy as np
import pandas as pd 

#美國球賽門票收入
tips_df = pd.read_csv('tips.csv')
tips_df.columns = ['總票價', '小費', '吸煙者', '日期', '時間', '大小']
tips_df['小費比例'] = tips_df['小費'] / tips_df['總票價'] #計算出新的一欄
tips_df

總票價    小費  吸煙者    日期      時間  大小      小費比例
0    16.99  1.01   No   Sun  Dinner   2  0.059447
1    10.34  1.66   No   Sun  Dinner   3  0.160542
2    21.01  3.50   No   Sun  Dinner   3  0.166587
3    23.68  3.31   No   Sun  Dinner   2  0.139780
4    24.59  3.61   No   Sun  Dinner   4  0.146808
..     ...   ...  ...   ...     ...  ..       ...
239  29.03  5.92   No   Sat  Dinner   3  0.203927
240  27.18  2.00  Yes   Sat  Dinner   2  0.073584
241  22.67  2.00  Yes   Sat  Dinner   2  0.088222
242  17.82  1.75   No   Sat  Dinner   2  0.098204
243  18.78  3.00   No  Thur  Dinner   2  0.159744

[244 rows x 7 columns]

In [34]:
grouped = tips_df.groupby(by=['吸煙者','日期']) #合併吸煙者、合併日期
functions = [('數量','count'),('平均','mean'),('最大值','max')] #統計量
tips_df3 = grouped[['小費','總票價']].agg(functions) #計算小費和總票價的統計量
tips_df3

小費                  總票價                  
          數量        平均    最大值  數量         平均    最大值
吸煙者 日期                                             
No  Fri    4  2.812500   3.50   4  18.420000  22.75
    Sat   45  3.102889   9.00  45  19.661778  48.33
    Sun   57  3.167895   6.00  57  20.506667  48.17
    Thur  45  2.673778   6.70  45  17.113111  41.19
Yes Fri   15  2.714000   4.73  15  16.813333  40.17
    Sat   42  2.875476  10.00  42  21.276667  50.81
    Sun   19  3.516842   6.50  19  24.120000  45.35
    Thur  17  3.030000   5.00  17  19.190588  43.11

In [ ]:
tips_df3['小費'] #看全部
tips_df3['小費'][['數量','平均']] #只看數量和平均

數量        平均
吸煙者 日期                
No  Fri    4  2.812500
    Sat   45  3.102889
    Sun   57  3.167895
    Thur  45  2.673778
Yes Fri   15  2.714000
    Sat   42  2.875476
    Sun   19  3.516842
    Thur  17  3.030000

In [ ]:
tips_df3.to_excel('tips.xlsx') #存成excel

render：https://dashboard.render.com/

- New→webservice→連結github
- Region：選Singapore（主機建立位置）
- Root Directory：根目錄（專案資料夾）
- Build Command：要安裝的套件
- Start Command：主程式碼
- Environment Variables：環境變數，有密碼才會用到